server/data/incidentTypeConfiguration/formWizard.ts

```diff
@@ -46,9 +48,168 @@ export function generateSteps(config: IncidentTypeConfiguration): FormWizard.Ste
       }
     })
 
+  groupSteps(steps)
+
   return steps
 }
 
+/**
+ * Merge all contiguous, non-branching steps in steps
+ *
+ * - If a step always lead to the same next step these can be potentially be
+ *   grouped.
+ * - Each step/question can only be in a single group.
+ *
+ * So steps with multiple parents (the right side of the branching)
+ * can't be the last step in a group. However these can be the
+ * beginning a new group.
+ *
+ * Example
+ * - Q-animals
+ *   => Q-dog
+ *   => Q-icecream
+ * - Q-dog
+ *   => Q-icecream
+ * - Q-icecream
+ *  => Q-end
+ * - Q-end
+ *  => null
+ *
+ * Here Q-animals branches and can lead to Q-dog or Q-icecream.
+ * This means q-animals would be in its own group.
+ * Both Q-animals and Q-dog lead to Q-icecream, so Q-icecream can't
+ * be part of these groups (each question belongs to a single group)
+ * Finally, Q-icecream always lead to Q-end, and Q-icecream is the only
+ * parent Q-end has, this means Q-end can be grouped into q-icecream
+ *
+ * @param steps to group
+ */
+function groupSteps(steps: FormWizard.Steps) {
+  const answersCounts: Map<string, number> = new Map()
+  const stepsWithSingleParent: Set<string | null> = buildStepsWithSingleParent()
+
+  /**
+   * Get the current number of answers for a given stepId
+   */
+  function getAnswersCountForStep(stepId: string) {
+    if (answersCounts.get(stepId) === undefined) {
+      answersCounts.set(stepId, countStepAnswers(stepId))
+    }
+    return answersCounts.get(stepId)
+  }
+
+  /**
+   * Count the number of answers for a given stepId
+   */
+  function countStepAnswers(stepId: string): number {
+    const step = steps[`/${stepId}`]
+    if (step === undefined) {
+      return 0
+    }
+
+    let count = 0
+    for (const nextStepCondition of Object.values(step.next as unknown as [{ value: string[] }])) {
```

Comment:
`step.next` is an array isn't it? so you don't need the `Object.values(…)` wrapper? not that it harms anything, it just threw me as i thought this was an array. (unless it's not lol)

```typescript
for (const nextStepCondition of (step.next as unknown as [{ value: string[] }])) { … }
```

Reply:
Correct, `step.next` is the array of `NextStepCondition` so you are right, `Object.values` is somewhat redundant.
Between JS weirdness and TS randomly complaining about things I must have played it safe plus it made it stick more explicit it was an array (as we also have `.next` `string | null` confusingly).

Lemme play with it, TS seems still somewhat happy with this:

```TypeScript
    for (const nextStepCondition of step.next) {
      count += (nextStepCondition as { value: string[] }).value.length
    }
```

which I think reads a bit better possibly?

Reply:
https://github.com/ministryofjustice/hmpps-incident-reporting/pull/252/commits/996af61f1844f8fce08443e416b1f949a64f3387 better 👍? worse 👎? I think it's slightly better but I still think not a fan of how we have to cast.

I suppose the challenge for TS with this kind of union types like `A | B | C` is that it doesn't know which one would it be. In this case is the following...

```TypeScript
{
          /** field, op and value. op defaults to '===' */
          field: string
          op?:
            | OperatorFunction
            | '>'
            | '>='
            | '<'
            | '<='
            | '=='
            | '==='
            | '!='
            | 'before'
            | 'after'
            | 'in'
            | 'all'
            | 'some'
          value: Value | MultiValue
          next: string | null
        }
```

I suppose we could have another type for this that we use in the definition of `NextStepCondition` but I think it would be just overcomplicating things possibly (and it may still not be enough because of the `Value | MultiValue`).

In this specific instance we only care about the `value` property being a `string[]` so possibly this is a bit more explicit.

Reply:
yeah totally. i mean in the absolute general case we'd have to do a whole bunch of `Array.isArray(step.next)` and `'value' in item && Array.isArray(item.value)`. this would make the code worse, but technically appease type checks